test/grouping.jl

```diff
@@ -636,77 +636,95 @@ end
 
     # Only test that different combine syntaxes work,
     # and rely on tests below for deeper checks
-    @test combine(:c => sum, gd) ==
-        combine(gd, :c => sum) ==
-        combine(gd, (:c => sum,)) ==
+    @test combine(gd, :c => sum) ==
+        combine(:c => sum, gd) ==
+        combine(gd, :c => sum => :c_sum) ==
+        combine(:c => sum => :c_sum, gd) ==
         combine(gd, [:c => sum]) ==
-        combine(gd, c_sum = :c => sum) ==
-        combine(:c => x -> (c_sum=sum(x),), gd) ==
-        combine(gd, :c => x -> (c_sum=sum(x),)) ==
+        combine(gd, [:c => sum => :c_sum]) ==
         combine(d -> (c_sum=sum(d.c),), gd) ==
         combine(gd, d -> (c_sum=sum(d.c),))
 
-    @test combine(:c => vexp, gd) ==
-        combine(gd, :c => vexp) ==
-        combine(gd, (:c => vexp,)) ==
+    @test combine(gd, :c => vexp) ==
+        combine(:c => vexp, gd) ==
+        combine(gd, :c => vexp => :c_function) ==
+        combine(:c => vexp => :c_function, gd) ==
         combine(gd, [:c => vexp]) ==
-        combine(gd, c_function = :c => vexp) ==
-        combine(:c => x -> (c_function=exp.(x),), gd) ==
-        combine(gd, :c => x -> (c_function=exp.(x),)) ==
+        combine(gd, [:c => vexp => :c_function]) ==
         combine(d -> (c_function=exp.(d.c),), gd) ==
         combine(gd, d -> (c_function=exp.(d.c),))
 
+
     @test combine(gd, :b => sum, :c => sum) ==
-        combine(gd, (:b => sum, :c => sum,)) ==
+        combine(gd, :b => sum => :b_sum, :c => sum => :c_sum) ==
         combine(gd, [:b => sum, :c => sum]) ==
-        combine(gd, b_sum = :b => sum, c_sum = :c => sum) ==
-        combine((:b, :c) => x -> (b_sum=sum(x.b), c_sum=sum(x.c)), gd) ==
-        combine(gd, (:b, :c) => x -> (b_sum=sum(x.b), c_sum=sum(x.c))) ==
+        combine(gd, [:b => sum => :b_sum, :c => sum => :c_sum]) ==
         combine(d -> (b_sum=sum(d.b), c_sum=sum(d.c)), gd) ==
         combine(gd, d -> (b_sum=sum(d.b), c_sum=sum(d.c)))
 
     @test combine(gd, :b => vexp, :c => identity) ==
-        combine(gd, (:b => vexp, :c => identity,)) ==
+        combine(gd, :b => vexp => :b_function, :c => identity => :c_identity) ==
         combine(gd, [:b => vexp, :c => identity]) ==
-        combine(gd, b_function = :b => vexp, c_identity = :c => identity) ==
-        combine((:b, :c) => x -> (b_function=vexp(x.b), c_identity=x.c), gd) ==
-        combine(gd, (:b, :c) => x -> (b_function=vexp(x.b), c_identity=x.c)) ==
+        combine(gd, [:b => vexp => :b_function, :c => identity => :c_identity]) ==
         combine(d -> (b_function=vexp(d.b), c_identity=d.c), gd) ==
-        combine(gd, d -> (b_function=vexp(d.b), c_identity=d.c))
+        combine(gd, d -> (b_function=vexp(d.b), c_identity=d.c)) ==
+        combine([:b, :c] => (b, c) -> (b_function=vexp(b), c_identity=c), gd) ==
+        combine(gd, [:b, :c] => (b, c) -> (b_function=vexp(b), c_identity=c))
+
+    @test by(x -> extrema(x.c), df, :a) == by(:c => (x -> extrema(x)) => :x1, df, :a)
```

Comment:
Why do you need to add these new tests here? I think the logic of the file was that all `by` variants are tested above against `combine`, and below we test that `combine` gives correct results.

Reply:
right - these are "variant" tests. So I changed them here to `combine` and added the equivalent tests for `by` above.